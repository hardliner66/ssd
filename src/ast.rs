#![allow(dead_code)]

use std::fmt::{Debug, Formatter};

#[derive(Debug, Clone)]
pub struct SsdcFile {
    pub namespace: Namespace,
    pub imports: Vec<Import>,
    pub data_types: Vec<DataType>,
    pub services: Vec<Service>,
}

const INDENT: &str = "    ";

impl SsdcFile {
    pub fn new(
        namespace: Namespace,
        imports: Vec<Import>,
        data_types: Vec<DataType>,
        services: Vec<Service>,
    ) -> Self {
        Self {
            namespace,
            imports,
            data_types,
            services,
        }
    }

    pub fn namespace(&mut self) -> Namespace {
        self.namespace.clone()
    }

    pub fn imports(&mut self) -> Vec<Import> {
        self.imports.clone()
    }

    pub fn data_types(&mut self) -> Vec<DataType> {
        self.data_types.clone()
    }

    pub fn services(&mut self) -> Vec<Service> {
        self.services.clone()
    }
}

#[derive(Debug, Clone)]
pub struct Import {
    pub path: Namespace,
    pub attributes: Vec<Attribute>,
}

impl Import {
    pub fn new(path: Namespace, attributes: Vec<Attribute>) -> Self {
        Import { path, attributes }
    }

    pub fn path(&mut self) -> Namespace {
        self.path.clone()
    }

    pub fn attributes(&mut self) -> Vec<Attribute> {
        self.attributes.clone()
    }
}

#[derive(Debug, Clone)]
pub struct Dependency {
    pub name: Namespace,
    pub attributes: Vec<Attribute>,
}

impl Dependency {
    pub fn new(name: Namespace, attributes: Vec<Attribute>) -> Self {
        Dependency { name, attributes }
    }

    pub fn name(&mut self) -> Namespace {
        self.name.clone()
    }

    pub fn attributes(&mut self) -> Vec<Attribute> {
        self.attributes.clone()
    }
}

impl ToString for Dependency {
    fn to_string(&self) -> String {
        self.name.to_string()
    }
}

fn format_attributes(v: &Vec<Attribute>, suffix: Option<&str>) -> String {
    if v.is_empty() {
        "".to_string()
    } else {
        format!(
            "@[{}]{}",
            v.iter()
                .map(|a| a.to_string())
                .collect::<Vec<_>>()
                .join(", "),
            suffix.unwrap_or_default()
        )
    }
}

impl ToString for SsdcFile {
    fn to_string(&self) -> String {
        format!(
            "{}\n\n{}\n\n{}",
            self.imports
                .iter()
                .map(|i| format!(
                    "{}import {};",
                    format_attributes(&i.attributes, Some("\n")),
                    i.path.to_string()
                ))
                .collect::<Vec<_>>()
                .join("\n"),
            self.data_types
                .iter()
                .map(|d| format!(
                    "{}{}",
                    format_attributes(&d.attributes, Some("\n")),
                    d.to_string()
                ))
                .collect::<Vec<_>>()
                .join("\n"),
            self.services
                .iter()
                .map(|s| format!(
                    "{}{}",
                    format_attributes(&s.attributes, Some("\n")),
                    s.to_string()
                ))
                .collect::<Vec<_>>()
                .join("\n"),
        )
    }
}

#[derive(Debug, Clone)]
pub struct Parameter {
    pub name: String,
    pub value: Option<String>,
}

impl Parameter {
    pub fn name(&mut self) -> String {
        self.name.clone()
    }

    pub fn value(&mut self) -> Option<String> {
        self.value.clone()
    }
}

#[derive(Debug, Clone)]
pub struct Attribute {
    pub name: Namespace,
    pub parameters: Vec<Parameter>,
}

impl Attribute {
    pub fn new(name: Namespace, parameters: Vec<(String, Option<String>)>) -> Self {
        Self {
            name,
            parameters: parameters
                .into_iter()
                .map(|(name, value)| Parameter { name, value })
                .collect(),
        }
    }

    pub fn name(&mut self) -> Namespace {
        self.name.clone()
    }

    pub fn parameters(&mut self) -> Vec<Parameter> {
        self.parameters.clone()
    }
}

impl ToString for Attribute {
    fn to_string(&self) -> String {
        if self.parameters.is_empty() {
            format!("{}", self.name.to_string())
        } else {
            format!(
                "{}({})",
                self.name.to_string(),
                self.parameters
                    .iter()
                    .map(|p| format!(
                        "{}{}",
                        p.name,
                        p.value
                            .clone()
                            .map(|v| format!(" = {v}"))
                            .unwrap_or_default()
                    ))
                    .collect::<Vec<_>>()
                    .join(", ")
            )
        }
    }
}

#[derive(Debug, Clone)]
pub struct DataType {
    pub name: String,
    pub properties: Vec<NameTypePair>,
    pub attributes: Vec<Attribute>,
}

impl DataType {
    pub fn new(name: String, properties: Vec<NameTypePair>, attributes: Vec<Attribute>) -> Self {
        Self {
            name,
            properties,
            attributes,
        }
    }

    pub fn name(&mut self) -> String {
        self.name.clone()
    }

    pub fn properties(&mut self) -> Vec<NameTypePair> {
        self.properties.clone()
    }

    pub fn attributes(&mut self) -> Vec<Attribute> {
        self.attributes.clone()
    }
}

impl ToString for DataType {
    fn to_string(&self) -> String {
        format!(
            "type {} {{\n{}\n}};\n",
            self.name,
            self.properties
                .iter()
                .map(|p| format!(
                    "{}{}{}: {},",
                    INDENT,
                    format_attributes(&p.attributes, Some("\n    ")),
                    p.name,
                    p.typ.to_string()
                ))
                .collect::<Vec<_>>()
                .join("\n"),
        )
    }
}

#[derive(Debug, Clone)]
pub struct Service {
    pub name: String,
    pub dependencies: Vec<Dependency>,
    pub handlers: Vec<Handler>,
    pub attributes: Vec<Attribute>,
}

impl Service {
    pub fn new(
        name: String,
        dependencies: Vec<Dependency>,
        handlers: Vec<Handler>,
        attributes: Vec<Attribute>,
    ) -> Self {
        Self {
            name,
            dependencies,
            handlers,
            attributes,
        }
    }

    pub fn name(&mut self) -> String {
        self.name.clone()
    }

    pub fn dependencies(&mut self) -> Vec<Dependency> {
        self.dependencies.clone()
    }

    pub fn handlers(&mut self) -> Vec<Handler> {
        self.handlers.clone()
    }

    pub fn attributes(&mut self) -> Vec<Attribute> {
        self.attributes.clone()
    }
}

impl ToString for Service {
    fn to_string(&self) -> String {
        format!(
            "service {} {{\n{}\n\n{}\n}};",
            self.name,
            &self
                .dependencies
                .iter()
                .map(|d| format!(
                    "{}{}depends on {};",
                    INDENT,
                    format_attributes(&d.attributes, Some(&format!("\n{}", INDENT))),
                    d.to_string()
                ))
                .collect::<Vec<_>>()
                .join("\n"),
            &self
                .handlers
                .iter()
                .map(|h| format!(
                    "{}{}{}",
                    INDENT,
                    format_attributes(&h.attributes, Some(&format!("\n{}", INDENT))),
                    h.to_string()
                ))
                .collect::<Vec<_>>()
                .join("\n"),
        )
    }
}

#[derive(Debug, Clone)]
pub struct Handler {
    pub name: String,
    pub arguments: Vec<NameTypePair>,
    pub return_type: Option<String>,
    pub attributes: Vec<Attribute>,
}

impl ToString for Handler {
    fn to_string(&self) -> String {
        format!(
            "handles {}({}){};",
            self.name,
            self.arguments
                .iter()
                .map(|a| format!(
                    "{}{}: {}",
                    format_attributes(&a.attributes, Some(" ")),
                    a.name,
                    a.typ.to_string(),
                ))
                .collect::<Vec<_>>()
                .join(", "),
            self.return_type
                .as_ref()
                .map(|t| format!(" -> {}", t))
                .unwrap_or(String::new())
        )
    }
}

impl Handler {
    pub fn new(
        name: String,
        arguments: Vec<NameTypePair>,
        return_type: Option<String>,
        attributes: Vec<Attribute>,
    ) -> Self {
        Self {
            name,
            arguments,
            return_type,
            attributes,
        }
    }

    pub fn name(&mut self) -> String {
        self.name.clone()
    }

    pub fn arguments(&mut self) -> Vec<NameTypePair> {
        self.arguments.clone()
    }

    pub fn return_type(&mut self) -> Option<String> {
        self.return_type.clone()
    }

    pub fn attributes(&mut self) -> Vec<Attribute> {
        self.attributes.clone()
    }
}

#[derive(Clone)]
pub struct NameTypePair {
    pub name: String,
    pub typ: Namespace,
    pub attributes: Vec<Attribute>,
}

impl Debug for NameTypePair {
    fn fmt(&self, fmt: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
        if fmt.alternate() {
            write!(
                fmt,
                "{}{}: {}",
                format_attributes(&self.attributes, None),
                self.name,
                self.typ.to_string()
            )
        } else {
            write!(
                fmt,
                "{{ name: {}, type: {:?}, attributes: {:?} }}",
                self.name, self.typ, self.attributes
            )
        }
    }
}

impl NameTypePair {
    pub fn new(name: String, typ: Namespace, attributes: Vec<Attribute>) -> Self {
        Self {
            name,
            typ,
            attributes,
        }
    }

    pub fn name(&mut self) -> String {
        self.name.clone()
    }

    pub fn typ(&mut self) -> Namespace {
        self.typ.clone()
    }

    pub fn attributes(&mut self) -> Vec<Attribute> {
        self.attributes.clone()
    }
}

#[derive(Debug, Clone)]
pub struct Namespace {
    components: Vec<String>,
}

impl IntoIterator for Namespace {
    type Item = String;

    type IntoIter = std::vec::IntoIter<Self::Item>;

    fn into_iter(self) -> Self::IntoIter {
        self.components.into_iter()
    }
}

impl Namespace {
    pub fn new(v: String) -> Self {
        Namespace {
            components: v.split("::").map(ToOwned::to_owned).collect(),
        }
    }

    pub fn from_vec(components: Vec<String>) -> Self {
        Namespace { components }
    }

    pub fn components(&mut self) -> Vec<String> {
        self.components.clone()
    }
}

impl ToString for Namespace {
    fn to_string(&self) -> String {
        self.components.join("::")
    }
}
