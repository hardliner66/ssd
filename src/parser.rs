use std::{num::ParseIntError, path::PathBuf};

use pest::{
    iterators::{Pair, Pairs},
    Parser, Span,
};
use pest_derive::Parser;
use regex::Regex;

use crate::ast::{
    Attribute, DataType, Dependency, Enum, EnumValue, Handler, Import, NameTypePair, Namespace,
    Service, SsdcFile,
};

fn parse_attribute_arg(node: Pair<Rule>) -> Result<(String, Option<String>), ParseError> {
    let span = node.as_span();
    let mut p = node.into_inner();
    let name = p
        .next()
        .ok_or_else(|| ParseError::new(ParseErrorType::IncompleteAttributeArg, span))?
        .as_str()
        .to_string();
    let value = p.next().map(|p| p.into_inner().as_str().to_string());
    Ok((name, value))
}

fn parse_attribute(node: Pair<Rule>) -> Result<Attribute, ParseError> {
    let span = node.as_span();
    let mut p = node.into_inner();
    let name = p.next();
    let mut args = Vec::new();
    for p in p {
        args.push(parse_attribute_arg(p)?);
    }
    Ok(Attribute::new(
        Namespace::new(
            name.ok_or_else(|| ParseError::new(ParseErrorType::IncompleteAttribute, span))?
                .as_str(),
        ),
        args,
    ))
}

fn parse_attributes(node: Pair<Rule>) -> Result<Vec<Attribute>, ParseError> {
    node.into_inner().map(parse_attribute).collect()
}

fn parse_name(p: &mut Pairs<Rule>, n: Pair<Rule>) -> Result<(String, Vec<Attribute>), ParseError> {
    let span = n.as_span();
    if n.as_rule() == Rule::attributes {
        let attributes = parse_attributes(n)?;
        let name = p
            .next()
            .ok_or_else(|| ParseError::new(ParseErrorType::IncompleteName, span))?
            .as_str()
            .to_string();
        Ok((name, attributes))
    } else {
        let name = n.as_str().to_string();
        Ok((name, Vec::new()))
    }
}

#[derive(Parser)]
#[grammar = "grammar.pest"]
pub(crate) struct FileParser;

#[derive(Debug)]
pub struct ParseError {
    pub error_type: ParseErrorType,
    pub span: String,
}

impl ParseError {
    fn new(error_type: ParseErrorType, span: Span) -> Self {
        Self {
            error_type,
            span: format!("{:?}", span),
        }
    }
}

#[derive(Debug)]
pub enum ParseErrorType {
    IncompleteImport,
    IncompleteDatatype,
    IncompleteProperty,
    MissingType(String),
    IncompleteEnum,
    IncompleteEnumValue,
    InvalidEnumValue(String),
    IncompleteService,
    IncompleteDepends,
    IncompleteHandler,
    IncompleteArgumentIdent,
    IncompleteAttributeArg,
    IncompleteAttribute,
    IncompleteName,
    UnexpectedElement(String),
    OtherError(String),
}

impl std::fmt::Display for ParseError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match &self.error_type {
            ParseErrorType::IncompleteImport => write!(f, "Import incomplete. ({})", self.span),
            ParseErrorType::IncompleteDatatype => write!(f, "Datatype incomplete. ({})", self.span),
            ParseErrorType::IncompleteProperty => write!(f, "Property incomplete. ({})", self.span),
            ParseErrorType::MissingType(name) => {
                write!(f, "Type missing after {}. ({:?})", name, self.span)
            }
            ParseErrorType::IncompleteService => write!(f, "Service incomplete. ({})", self.span),
            ParseErrorType::IncompleteDepends => write!(f, "Depends incomplete. ({})", self.span),
            ParseErrorType::IncompleteHandler => write!(f, "Handler incomplete. ({})", self.span),
            ParseErrorType::IncompleteArgumentIdent => {
                write!(f, "Argument ident incomplete. ({})", self.span)
            }
            ParseErrorType::IncompleteAttributeArg => {
                write!(f, "Attribute argument incomplete. ({})", self.span)
            }
            ParseErrorType::IncompleteAttribute => {
                write!(f, "Attribute incomplete. ({})", self.span)
            }
            ParseErrorType::IncompleteName => {
                write!(f, "Name incomplete. ({})", self.span)
            }
            ParseErrorType::UnexpectedElement(info) => {
                write!(f, "Unexpected element {} ({})", info, self.span)
            }
            ParseErrorType::IncompleteEnum => write!(f, "Incomplete enum. ({})", self.span),
            ParseErrorType::IncompleteEnumValue => {
                write!(f, "Incomplete enum value. ({})", self.span)
            }
            ParseErrorType::InvalidEnumValue(info) => {
                write!(f, "Invalid enum value. {} ({})", info, self.span)
            }
            ParseErrorType::OtherError(inner) => {
                write!(f, "Other({})", inner)
            }
        }
    }
}

impl ParseError {
    fn from_dyn_error<T: std::error::Error>(err: T) -> Self {
        ParseError {
            error_type: ParseErrorType::OtherError(format!("{}", err)),
            span: String::new(),
        }
    }
}

impl std::error::Error for ParseError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        None
    }

    fn cause(&self) -> Option<&dyn std::error::Error> {
        self.source()
    }
}

#[allow(clippy::too_many_lines)]
pub fn parse(content: &str, namespace: Namespace) -> Result<SsdcFile, ParseError> {
    use ParseErrorType::{
        IncompleteArgumentIdent, IncompleteDatatype, IncompleteDepends, IncompleteEnum,
        IncompleteEnumValue, IncompleteHandler, IncompleteImport, IncompleteProperty,
        IncompleteService, InvalidEnumValue, MissingType, UnexpectedElement,
    };
    let pairs = FileParser::parse(Rule::file, content).map_err(ParseError::from_dyn_error)?;
    let mut imports = Vec::new();
    let mut datatypes = Vec::new();
    let mut enums = Vec::new();
    let mut services = Vec::new();

    for p in pairs {
        match p.as_rule() {
            Rule::import => {
                let span = p.as_span();
                let mut p = p.into_inner();
                let n = p
                    .next()
                    .ok_or_else(|| ParseError::new(IncompleteImport, span))?;
                let (name, attributes) = parse_name(&mut p, n)?;
                imports.push(Import::new(Namespace::new(&name), attributes));
            }
            Rule::data => {
                let span = p.as_span();
                let mut p = p.into_inner();
                let n = p
                    .next()
                    .ok_or_else(|| ParseError::new(IncompleteDatatype, span))?;
                let (name, attributes) = parse_name(&mut p, n)?;

                let mut properties = Vec::new();

                for p in p {
                    let span = p.as_span();
                    let mut p = p.into_inner();
                    let n = p
                        .next()
                        .ok_or_else(|| ParseError::new(IncompleteProperty, span))?;
                    let (name, attributes) = parse_name(&mut p, n)?;
                    let typ = p
                        .next()
                        .ok_or_else(|| ParseError::new(MissingType(name.clone()), span))?
                        .as_str()
                        .to_string();
                    properties.push(NameTypePair::new(name, Namespace::new(&typ), attributes));
                }

                datatypes.push(DataType::new(name, properties, attributes));
            }
            Rule::enum_ => {
                let span = p.as_span();
                let mut p = p.into_inner();
                let n = p
                    .next()
                    .ok_or_else(|| ParseError::new(IncompleteEnum, span))?;
                let (name, attributes) = parse_name(&mut p, n)?;

                let mut values = Vec::new();

                for p in p {
                    let span = p.as_span();
                    let mut p = p.into_inner();
                    let n = p
                        .next()
                        .ok_or_else(|| ParseError::new(IncompleteEnumValue, span))?;
                    let (name, attributes) = parse_name(&mut p, n)?;
                    let value = if let Some(v) = p.next() {
                        Some(v.as_str().parse().map_err(|err: ParseIntError| {
                            ParseError::new(InvalidEnumValue(err.to_string()), span)
                        })?)
                    } else {
                        None
                    };
                    values.push(EnumValue::new(name, value, attributes));
                }

                enums.push(Enum::new(name, values, attributes));
            }
            Rule::service => {
                let span = p.as_span();
                let mut p = p.into_inner();
                let n = p
                    .next()
                    .ok_or_else(|| ParseError::new(IncompleteService, span))?;
                let (service_name, attributes) = parse_name(&mut p, n)?;

                let mut dependencies = Vec::new();
                let mut handlers = Vec::new();

                for p in p {
                    match p.as_rule() {
                        Rule::depends => {
                            let span = p.as_span();
                            let mut p = p.into_inner();
                            let n = p
                                .next()
                                .ok_or_else(|| ParseError::new(IncompleteDepends, span))?;
                            let (name, attributes) = parse_name(&mut p, n)?;
                            dependencies.push(Dependency::new(Namespace::new(&name), attributes));
                        }
                        Rule::handler => {
                            let span = p.as_span();
                            let mut p = p.into_inner();
                            let n = p
                                .next()
                                .ok_or_else(|| ParseError::new(IncompleteHandler, span))?;
                            let (handler_name, handler_attributes) = parse_name(&mut p, n)?;
                            let mut arguments = Vec::new();
                            let mut return_type = None;
                            let mut attributes = Vec::new();
                            for p in p.by_ref() {
                                match p.as_rule() {
                                    Rule::argument => {
                                        let span = p.as_span();
                                        let mut p = p.clone().into_inner();
                                        while let Some(n) = p.next() {
                                            match n.as_rule() {
                                                Rule::ident => {
                                                    let name = n.as_str().to_string();
                                                    let typ = p.next().ok_or_else(|| ParseError::new(IncompleteArgumentIdent, span))?.as_str().to_string();
                                                    arguments.push(NameTypePair::new(name, Namespace::new(&typ), attributes.clone()));
                                                    attributes.clear();
                                                }
                                                Rule::attributes => {
                                                    attributes = parse_attributes(n)?;
                                                }
                                                _ => Err(ParseError::new(
                                                    UnexpectedElement(format!(
                                                        "while parsing argument for handler \"{}\" in service \"{}\"! {}",
                                                        handler_name,service_name, p
                                                    )),
                                                    span,
                                                ))?,
                                            }
                                        }
                                    }
                                    Rule::typ => {
                                        let re = Regex::new(r"\s+").expect("invalid regex");
                                        return_type =
                                            Some(Namespace::new(&re.replace_all(p.as_str(), " ")));
                                    }
                                    _ => Err(ParseError::new(
                                        UnexpectedElement(format!(
                                            "while parsing handler \"{}\" in service \"{}\"! {}",
                                            handler_name, service_name, p
                                        )),
                                        p.as_span(),
                                    ))?,
                                }
                            }

                            if let Some(p) = p.next() {
                                if p.as_rule() == Rule::typ {
                                    return_type = Some(Namespace::new(p.as_str()));
                                } else {
                                    Err(ParseError::new(
                                        UnexpectedElement(format!(
                                            "while parsing return type for handler \"{}\" in service \"{}\"! {}",
                                            handler_name,service_name, p
                                        )),
                                        p.as_span(),
                                    ))?;
                                }
                            }
                            handlers.push(Handler::new(
                                handler_name,
                                arguments,
                                return_type,
                                handler_attributes,
                            ));
                        }
                        _ => Err(ParseError::new(
                            UnexpectedElement(format!(
                                "while parsing service \"{}\"! {}",
                                service_name, p
                            )),
                            p.as_span(),
                        ))?,
                    }
                }

                services.push(Service::new(
                    service_name,
                    dependencies,
                    handlers,
                    attributes,
                ));
            }
            Rule::EOI => {}
            _ => Err(ParseError::new(
                UnexpectedElement(format!("{}", p)),
                p.as_span(),
            ))?,
        }
    }

    Ok(SsdcFile::new(
        namespace, imports, datatypes, enums, services,
    ))
}

pub fn parse_file(base: &PathBuf, path: PathBuf) -> Result<SsdcFile, ParseError> {
    let content = std::fs::read_to_string(&path).map_err(ParseError::from_dyn_error)?;

    let mut path = if path.starts_with(base) {
        path.strip_prefix(base)
            .map_err(ParseError::from_dyn_error)?
            .to_owned()
    } else {
        path
    };

    path.set_extension("");
    let components = path
        .components()
        .map(|c| c.as_os_str().to_string_lossy().to_string())
        .collect::<Vec<_>>();

    parse(&content, Namespace::from_vec(components))
}

#[test]
fn test_simple() {
    insta::assert_ron_snapshot!(parse(
        include_str!("../data/test.svc"),
        Namespace::new("__test__")
    )
    .unwrap());
}
