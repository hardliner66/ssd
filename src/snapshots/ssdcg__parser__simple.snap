---
source: src/parser.rs
expression: "parse(include_str!(\"../data/test.svc\"), Namespace::new(\"__test__\")).unwrap()"
---
SsdcFile(
  namespace: Namespace(
    components: [
      "__test__",
    ],
  ),
  imports: [
    Import(
      path: Namespace(
        components: [
          "std",
          "path",
          "Path",
        ],
      ),
      attributes: [],
    ),
    Import(
      path: Namespace(
        components: [
          "std",
          "error",
          "Error",
        ],
      ),
      attributes: [
        Attribute(
          name: Namespace(
            components: [
              "test",
            ],
          ),
          parameters: [],
        ),
        Attribute(
          name: Namespace(
            components: [
              "abcd",
            ],
          ),
          parameters: [
            Parameter(
              name: "foo",
              value: None,
            ),
            Parameter(
              name: "bar",
              value: Some("\"blub\""),
            ),
          ],
        ),
      ],
    ),
  ],
  data_types: [
    DataType(
      name: "Rect",
      properties: [
        NameTypePair(
          name: "x",
          typ: Namespace(
            components: [
              "i32",
            ],
          ),
          attributes: [
            Attribute(
              name: Namespace(
                components: [
                  "test",
                ],
              ),
              parameters: [],
            ),
          ],
        ),
        NameTypePair(
          name: "y",
          typ: Namespace(
            components: [
              "i32",
            ],
          ),
          attributes: [],
        ),
        NameTypePair(
          name: "width",
          typ: Namespace(
            components: [
              "i32",
            ],
          ),
          attributes: [],
        ),
        NameTypePair(
          name: "height",
          typ: Namespace(
            components: [
              "i32",
            ],
          ),
          attributes: [],
        ),
      ],
      attributes: [
        Attribute(
          name: Namespace(
            components: [
              "test",
            ],
          ),
          parameters: [],
        ),
      ],
    ),
  ],
  services: [
    Service(
      name: "Abcd",
      dependencies: [
        Dependency(
          name: Namespace(
            components: [
              "some",
              "other",
              "Service",
            ],
          ),
          attributes: [
            Attribute(
              name: Namespace(
                components: [
                  "test",
                ],
              ),
              parameters: [],
            ),
          ],
        ),
      ],
      handlers: [
        Handler(
          name: "test",
          arguments: [
            NameTypePair(
              name: "i",
              typ: Namespace(
                components: [
                  "i32",
                ],
              ),
              attributes: [
                Attribute(
                  name: Namespace(
                    components: [
                      "whatever",
                    ],
                  ),
                  parameters: [],
                ),
              ],
            ),
            NameTypePair(
              name: "u",
              typ: Namespace(
                components: [
                  "u32",
                ],
              ),
              attributes: [],
            ),
          ],
          return_type: Some(Namespace(
            components: [
              "String",
            ],
          )),
          attributes: [
            Attribute(
              name: Namespace(
                components: [
                  "test",
                ],
              ),
              parameters: [],
            ),
          ],
        ),
        Handler(
          name: "test",
          arguments: [
            NameTypePair(
              name: "i",
              typ: Namespace(
                components: [
                  "i32",
                ],
              ),
              attributes: [
                Attribute(
                  name: Namespace(
                    components: [
                      "whatever",
                    ],
                  ),
                  parameters: [],
                ),
              ],
            ),
            NameTypePair(
              name: "f",
              typ: Namespace(
                components: [
                  "f32",
                ],
              ),
              attributes: [],
            ),
          ],
          return_type: Some(Namespace(
            components: [
              "String",
            ],
          )),
          attributes: [],
        ),
        Handler(
          name: "read_file",
          arguments: [
            NameTypePair(
              name: "path",
              typ: Namespace(
                components: [
                  "Path",
                ],
              ),
              attributes: [],
            ),
          ],
          return_type: Some(Namespace(
            components: [
              "String",
            ],
          )),
          attributes: [
            Attribute(
              name: Namespace(
                components: [
                  "throws",
                ],
              ),
              parameters: [
                Parameter(
                  name: "Error",
                  value: None,
                ),
              ],
            ),
          ],
        ),
      ],
      attributes: [
        Attribute(
          name: Namespace(
            components: [
              "test",
            ],
          ),
          parameters: [],
        ),
      ],
    ),
  ],
)
